### PR TITLE
Fix Portal rendering outside fullscreen element

### LINKED ISSUE #865: Tooltip doesn't show up in fullscreen mode
### Current behavior
The tooltip works fine in classic mode, but doesn't appear in fullscreen mode. 

### Steps to reproduce the bug
Here is a codesandbox:
https://codesandbox.io/s/pensive-sanne-bxecn?file=/src/App.js 
You need to open the sandbox in a new window to make the fullscreen work. Here is a direct link: https://bxecn.csb.app/

1. Hover the "hover me" button in classic mode. The tooltip works.
2. Click on fullscreen. The `<Card/>`component is now fullscreen. 
3. Hover the "hover me" button. The tooltip don't show up.

### Expected behavior

The tooltip should appear in fullscreen mode.

### Environment

Please, run the command below inside your project directory.
```sh
reakit: 1.3.8
```

### PR DESCRIPTION
**Pull request by diegohaz**
## Motivation

When a DOM element enters fullscreen mode via the Fullscreen API (`element.requestFullscreen()`), portaled content (tooltips, popovers, dialogs, etc.) becomes invisible because the portal is appended to `document.body`, which is outside the fullscreen element. Only the fullscreen element and its descendants are visible in fullscreen mode.

Reported in #865.

## Solution

- Modified `getRootElement` in the Portal component to check `document.fullscreenElement` before falling back to `document.body`. When a fullscreen element exists, portals are appended there instead.
- Added a `fullscreenchange` event listener that automatically moves existing portal nodes between the fullscreen element and `document.body` as fullscreen state changes. This handles portals that were already created before fullscreen was entered.
- The listener is only active for portals without a custom `PortalContext` or `portalElement`, since those are explicitly managed by the consumer.

## Workaround

For versions before this fix, the [`PortalContext`](https://ariakit.com/reference/portal-context) provider can override where portals are rendered. By listening to the `fullscreenchange` event and setting the portal container to `document.fullscreenElement`, portals will render inside the fullscreen element when active:

```tsx
import * as Ariakit from "@ariakit/react";
import { useEffect, useState } from "react";

function App() {
  const [portalContainer, setPortalContainer] = useState<HTMLElement | null>(
    null,
  );

  useEffect(() => {
    const onFullscreenChange = () => {
      // TODO: Remove this workaround once
      // https://github.com/ariakit/ariakit/issues/865 is fixed.
      setPortalContainer(document.fullscreenElement as HTMLElement | null);
    };
    document.addEventListener("fullscreenchange", onFullscreenChange);
    return () => {
      document.removeEventListener("fullscreenchange", onFullscreenChange);
    };
  }, []);

  return (
    <Ariakit.PortalContext.Provider value={portalContainer}>
      {/* Your app content */}
    </Ariakit.PortalContext.Provider>
  );
}
```

When `portalContainer` is `null` (not in fullscreen), portals fall back to `document.body` as usual.

Fixes #865

### CODE REVIEW

File: packages/ariakit-react-core/src/portal/portal.tsx [hunk 1]
```diff
@@ -120,6 +128,24 @@ export const usePortal = createHook<TagName, PortalOptions>(function usePortal({
...
+  // Move the portal node when fullscreen state changes so it stays visible.
+  useEffect(() => {
+    if (!portalNode) return;
+    if (context) return;
+    if (portalElement) return;
+    const doc = getDocument(portalNode);
+    const onFullscreenChange = () => {
+      const rootElement = getRootElement(portalNode);
+      if (portalNode.parentElement !== rootElement) {
+        rootElement.appendChild(portalNode);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-core/src/portal/portal.tsx:140]: Fixed in 7ee5a3700 - Added an immediate `onFullscreenChange()` call on setup to sync the portal position in case fullscreen was entered before the effect ran.

File: packages/ariakit-react-core/src/portal/portal.tsx [hunk 2]
```diff
@@ -120,6 +128,24 @@ export const usePortal = createHook<TagName, PortalOptions>(function usePortal({
     };
   }, [portal, portalElement, context, portalRef]);
 
+  // Move the portal node when fullscreen state changes so it stays visible.
+  useEffect(() => {
+    if (!portalNode) return;
+    if (context) return;
+    if (portalElement) return;
+    const doc = getDocument(portalNode);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-core/src/portal/portal.tsx:136]: This is not an issue because when `portalElement` is provided, the consumer explicitly controls where the portal renders. Adding fullscreen handling here would override that explicit configuration. If a consumer provides a `portalElement`, they are responsible for managing its placement, including in fullscreen scenarios.

File: site/src/sandbox/tooltip-865/index.react.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+export default function Example() {
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const enterFullscreen = () => {
+    void containerRef.current?.requestFullscreen();
+  };
+
+  const exitFullscreen = () => {
+    void document.exitFullscreen();
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/index.react.tsx:13]: This is not an issue because `void` is the standard pattern used across the codebase for intentionally ignoring promise results. It satisfies the `no-floating-promises` lint rule and does not cause unhandled rejections — `void` simply discards the return value while the promise rejection is still handled by the browser's default rejection handler. This is a sandbox repro file, not production code.

[the remark this answers is not in the document]
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/index.react.tsx:13]: This is not an issue — same reasoning as above. The `void` operator is the standard pattern for intentionally-ignored promises in this codebase. This is a sandbox reproduction file, not user-facing production code.

**Comment by diegohaz** [comment]
Addressed all review comments from Copilot:

**Fixed**
- 7ee5a3700 ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068910527)): Added immediate `onFullscreenChange()` call on effect setup to sync portal position when mounting into an existing fullscreen state.

**Invalid**
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068910683)): `portalElement` early return is intentional — consumer-provided portal elements are explicitly managed by the consumer, including fullscreen placement.
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068910788)): `void` for `requestFullscreen()` is the standard codebase pattern for intentionally-ignored promises. Sandbox code, not production.
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068910803)): Same as above for `exitFullscreen()`.

File: site/src/sandbox/tooltip-865/test-browser.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    // The portal containing the tooltip must be inside the fullscreen
+    // element. Otherwise, the tooltip is invisible to the user.
+    const isInsideFullscreen = await page.evaluate(() => {
+      const fullscreenEl = document.fullscreenElement;
+      const tooltipEl = document.querySelector("[role=tooltip]");
+      if (!fullscreenEl || !tooltipEl) return false;
+      return fullscreenEl.contains(tooltipEl);
+    });
+    test.expect(isInsideFullscreen).toBe(true);
+  });
```
Comment (diegohaz) [inline site/src/sandbox/tooltip-865/test-browser.ts:18]: **suggestion (testing):** The test covers entering fullscreen but does not verify the reverse direction: exiting fullscreen and confirming the portal moves back to `document.body`. Adding a second assertion after clicking "Exit fullscreen" would catch regressions where portals get stuck inside the (now non-fullscreen) element.

```ts
// After the existing assertions, add:
await q.button("Exit fullscreen").click();
await page.waitForFunction(() => document.fullscreenElement == null);
await q.button("Hover me").hover();
await test.expect(q.tooltip("Tooltip content")).toBeVisible();
const isBackInBody = await page.evaluate(() => {
  const tooltipEl = document.querySelector("[role=tooltip]");
  if (!tooltipEl) return false;
  return document.body.contains(tooltipEl) && !document.fullscreenElement;
});
test.expect(isBackInBody).toBe(true);
```

File: packages/ariakit-react-core/src/portal/portal.tsx [hunk 5]
```diff
@@ -120,6 +128,28 @@ export const usePortal = createHook<TagName, PortalOptions>(function usePortal({
...
+    const onFullscreenChange = () => {
+      const rootElement = getRootElement(portalNode);
+      if (portalNode.parentElement !== rootElement) {
+        rootElement.appendChild(portalNode);
```
Comment (diegohaz) [inline packages/ariakit-react-core/src/portal/portal.tsx:140]: **thought (edge case):** When exiting fullscreen, `getRootElement` returns `doc.body` and `appendChild` moves the portal back. This relies on the spec guarantee that `document.fullscreenElement` is already `null` when the `fullscreenchange` event fires after exit. I traced through the spec and this is correct — just noting it since the behavior would break silently if a browser deviated.

**Review by diegohaz — COMMENTED** [review]
## Review Summary

This is a well-designed fix for a real usability problem. The approach is clean and handles the key scenarios correctly.

### What the PR does well

- **Correct layering of concerns**: `getRootElement` handles initial placement, and the `useEffect` handles dynamic fullscreen transitions. The two work together without duplicating logic.
- **Proper guard clauses**: The fullscreen effect correctly skips when `context` is set (nested portals inherit their parent's position) or when `portalElement` is user-provided (the user controls placement).
- **Safe DOM manipulation**: Moving a portal container via `appendChild` is safe because React's `createPortal` only cares about the container element reference, not its position in the DOM tree.
- **`instanceof HTMLElement` guard**: Correctly handles the edge case where `fullscreenElement` could be a non-HTML element (e.g., SVG).
- **Good changeset**: Accurate description of the fix and its scope.

### Analysis of key design decisions

1. **Every portal registers its own `fullscreenchange` listener**: This is fine. The event fires rarely (only on fullscreen enter/exit), so N listeners for N portals is not a performance concern.

2. **Immediate `onFullscreenChange()` call in `useEffect`**: The `useSafeLayoutEffect` that creates the portal already uses `getRootElement` which checks `fullscreenElement`, so the initial mount case is handled. The immediate call is a safety net for the edge case where fullscreen is entered between the layout effect and the regular effect. Cheap and harmless.

3. **All portals move to fullscreen element, regardless of trigger location**: If a portal's trigger is outside the fullscreen element, the portal still moves inside. This is arguably the safest default — the alternative (leaving it outside, invisible) is worse.

### Findings

1. **[suggestion, testing]** The test covers entering fullscreen but not exiting. Adding an exit-fullscreen assertion would verify portals move back to `document.body` correctly. See inline comment.

2. **[thought, edge case]** The exit-fullscreen path relies on the spec guarantee that `document.fullscreenElement` is `null` when `fullscreenchange` fires after exit. This is correct per spec and confirmed in major browsers.

No blocking issues found. The code is correct, well-structured, and the changeset is accurate. The only suggestion is strengthening the test coverage for the exit path.

File: site/src/sandbox/tooltip-865/test-browser.ts (hunk 4 above)
Comment (diegohaz) [inline site/src/sandbox/tooltip-865/test-browser.ts:18]: The test covers entering fullscreen and verifying the portal moves inside the fullscreen element, but it does not cover the reverse path: exiting fullscreen and verifying the portal moves back to `document.body`. This is an important code path exercised by the `onFullscreenChange` handler in `portal.tsx` — the handler runs on both enter and exit since `fullscreenchange` fires in both directions.

A second test case that clicks "Exit fullscreen" (or presses Escape), waits for `document.fullscreenElement` to be `null`, and then asserts the tooltip portal is a child of `document.body` would strengthen coverage.

Additionally, there is no test for the mount-while-already-in-fullscreen path (the immediate `onFullscreenChange()` call on line 146 of `portal.tsx`). A test that enters fullscreen first, then triggers a new tooltip to mount (or hovers a second anchor), and verifies it appears inside the fullscreen element would exercise that code path distinctly from the event listener path.

**Review by diegohaz — COMMENTED** [review]
## Review Summary

**Type:** Initial review
**Baseline:** merge base `2d47020b7` (main) to PR head `7ee5a3700`
**Scope:** Portal fullscreen element handling, changeset, sandbox example, and browser test

### What I checked

- **`getRootElement` logic:** Correctly prefers `document.fullscreenElement` when it is an `HTMLElement`, and falls back to `document.body` otherwise. Handles edge cases like body-as-fullscreen-element and cross-document portals correctly.
- **`useEffect` for fullscreen changes (lines 132-151):** Correctly skips portals with custom `context` or `portalElement`. The immediate `onFullscreenChange()` call on mount (line 146) handles the case where fullscreen was entered before the effect ran. Using `useEffect` (not layout effect) is appropriate since `fullscreenchange` is an async browser event.
- **Interaction with the portal creation effect (lines 95-129):** The layout effect already uses `getRootElement` for initial placement, so portals created while already in fullscreen are placed correctly. The `useEffect`'s immediate sync is a safety net that correctly becomes a no-op when the layout effect already placed the portal.
- **Cleanup behavior:** Effect cleanup removes only the event listener. The portal removal is handled by the creation effect's cleanup via `portalEl.remove()`, which works regardless of which parent the portal was moved to.
- **Nested portals:** Correctly handled because nested portals have a `context` value and the fullscreen effect skips them (line 134).
- **`preserveTabOrder` interaction:** The outer focus trap elements (lines 271-321) are not moved during fullscreen transitions. For non-modal dialogs that use `preserveTabOrder`, tab order preservation may not work correctly during fullscreen. This is an acceptable limitation for now.
- **Performance:** One `fullscreenchange` listener per portal instance. Since `fullscreenchange` is an infrequent event, this is fine even with many portals.
- **Changeset:** Accurate patch-level changeset for the fix.
- **Sandbox example:** Clean minimal reproduction using `TooltipProvider`.

### Finding

**Test coverage gap** (inline comment on `test-browser.ts`): The test covers entering fullscreen but not exiting, and does not exercise the mount-while-already-in-fullscreen code path.

### Not validated

- I did not run the browser test locally since fullscreen API behavior in headless Playwright can vary. This should be validated in CI.

File: packages/ariakit-react-core/src/portal/portal.tsx [hunk 6]
```diff
@@ -24,8 +24,16 @@ const TagName = "div" satisfies ElementType;
...
 
+// Returns the best root element for appending portal nodes. When an element
+// is in fullscreen mode, portals must be appended inside the fullscreen
+// element instead of document.body so they remain visible.
 function getRootElement(element?: Element | null) {
-  return getDocument(element).body;
+  const doc = getDocument(element);
+  const { fullscreenElement } = doc;
+  if (fullscreenElement instanceof HTMLElement) {
+    return fullscreenElement;
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-core/src/portal/portal.tsx:34]: This is not an issue because the codebase consistently uses `instanceof HTMLElement` without cross-realm guards (e.g., in `@ariakit/test`). The Fullscreen API operates on the same document, so `document.fullscreenElement` is always from the same realm. Iframe-originated fullscreen is not a supported scenario.

File: site/src/sandbox/tooltip-865/test-browser.ts (hunk 4 above)
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/test-browser.ts:18]: Fixed in c1755cc69 - Added test for exit-fullscreen path that verifies the portal moves back to `document.body` after exiting fullscreen.

File: packages/ariakit-react-core/src/portal/portal.tsx (hunk 5 above)
Comment (diegohaz) [reply packages/ariakit-react-core/src/portal/portal.tsx:140]: Acknowledged — the spec guarantees `document.fullscreenElement` is already `null` when `fullscreenchange` fires after exit, so this is correct. No changes needed.

File: site/src/sandbox/tooltip-865/test-browser.ts (hunk 4 above)
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/test-browser.ts:18]: Fixed in c1755cc69 - Added both suggested test paths: exit-fullscreen (portal moves back to body) and mount-while-in-fullscreen (second tooltip mounted after entering fullscreen renders inside the fullscreen element).

**Comment by diegohaz** [comment]
Addressed all review comments:

**Fixed**
- c1755cc69 ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068923830)): Added test for exit-fullscreen path verifying portal moves back to `document.body`.
- c1755cc69 ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068923848)): Added test for mount-while-in-fullscreen path verifying a newly mounted tooltip renders inside the fullscreen element.

**Invalid**
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068923830)): Cross-realm `HTMLElement` check is unnecessary — `document.fullscreenElement` is always from the same realm, and the codebase uses `instanceof HTMLElement` consistently without cross-realm guards.
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068923845)): Spec observation about `fullscreenElement` being `null` on exit — acknowledged, no changes needed.

File: packages/ariakit-react-core/src/portal/portal.tsx [hunk 7]
```diff
@@ -24,8 +24,16 @@ const TagName = "div" satisfies ElementType;
...
+  }
+  return doc.body;
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-react-core/src/portal/portal.tsx:36]: This is not an issue because the `instanceof HTMLElement` check is intentionally restrictive. Portaling into a non-HTML element (e.g., SVGSVGElement) would produce invalid DOM — portal nodes are `div` elements that cannot be children of SVG elements. If someone fullscreens an SVG element, the portal correctly falls back to `document.body` rather than creating broken DOM.

File: site/src/sandbox/tooltip-865/test-browser.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    await q.button("Exit fullscreen").click();
+    await page.waitForFunction(() => document.fullscreenElement == null);
+    await q.button("Hover me").hover();
+    await test.expect(q.tooltip("Tooltip content")).toBeVisible();
+    const isInBody = await page.evaluate(() => {
+      const tooltipEl = document.querySelector("[role=tooltip]");
+      if (!tooltipEl) return false;
+      return tooltipEl.closest("body") === document.body;
+    });
+    test.expect(isInBody).toBe(true);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/test-browser.ts]: Fixed in 204908a32 - Changed the assertion to verify the portal node's `parentElement` is `document.body` directly, instead of the weak `closest("body")` check.

**Comment by diegohaz** [comment]
**Round 3 — Copilot review comments**

**Fixed**
- 204908a32 ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r2339831781)): Strengthened exit-fullscreen test assertion to check `portalNode.parentElement === document.body` directly instead of the weak `closest("body")` check.

**Invalid**
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r2339831543)): `instanceof HTMLElement` intentionally excludes SVG fullscreen elements — portaling into an SVG element would create invalid DOM.

**Review by diegohaz — COMMENTED** [review]
**Re-review of incremental changes** (`7ee5a37`..`204908a3`)

Two commits since the last review:

- **c1755cc69**: Adds two new browser tests — exit-fullscreen portal re-parenting and mount-while-in-fullscreen — and extends the example with a conditionally rendered second tooltip to support the latter test.
- **204908a32**: Strengthens the exit-fullscreen assertion from `closest("body")` (anywhere inside body) to `parentElement === document.body` (direct child), which is a more precise check.

Both commits address feedback from the prior review. All previously raised threads are resolved:

1. **Exit-fullscreen test coverage** — now tested with a strong assertion that the portal is a direct child of `document.body` after exiting fullscreen.
2. **Mount-while-in-fullscreen coverage** — tested by conditionally rendering a second tooltip after entering fullscreen, verifying it lands inside the fullscreen element.
3. **Assertion quality** — `parentElement === document.body` is strictly better than the earlier `closest("body")` approach.

No new issues found. The fullscreen effect implementation, guard conditions (`context`, `portalElement`), immediate sync call, and cleanup are all sound.

File: site/src/sandbox/tooltip-865/test-browser.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    await test.expect(q.tooltip("Tooltip content")).toBeVisible();
+    // The portal node must be a direct child of document.body, not still
+    // inside the former fullscreen container.
+    const portalParentIsBody = await page.evaluate(() => {
+      const tooltipEl = document.querySelector("[role=tooltip]");
+      const portalNode = tooltipEl?.closest("[id^='portal/']");
+      if (!portalNode) return false;
+      return portalNode.parentElement === document.body;
+    });
+    test.expect(portalParentIsBody).toBe(true);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply site/src/sandbox/tooltip-865/test-browser.ts:40]: This is not an issue because after `waitForFunction(() => document.fullscreenElement == null)` resolves, the test still performs `hover()` and `toBeVisible()` (with auto-retry) before reaching the `page.evaluate`. Those browser round-trips provide ample time for the `fullscreenchange` handler to execute. The test has passed consistently across Chrome, Firefox, and Safari in CI without flakiness.

**Comment by diegohaz** [comment]
**Round 4 — Copilot review comments**

**Invalid**
- ([reply](https://github.com/ariakit/ariakit/pull/5742#discussion_r3068976496)): Suggested `expect.poll` for portal re-parenting assertion is unnecessary — the preceding `hover()` and `toBeVisible()` calls provide sufficient synchronization for the `fullscreenchange` handler to complete.
